### ISSUE TITLE
Should API v1.Secret support creating a secret from the contents of a file.

### ISSUE THREAD

**Issue by sijnc** [kubernetes/kubernetes#24744]
Currently AFAIK you can only create a secret from a file using:
```
kubectl create secret generic <name> --from-file=<key>=<file>
```

The drawback is this, If a change is made to <file> you need to delete and then recreate the secret:
```
kubectl delete secret <name>
kubectl create secret generic <name> --from-file=<key>=<file>
```

This is awkward because to update the other items (service, deployment, pv) in the same app you just run:
```
kubectl apply -f <filename>
```

I realize that I can create a script to base64 encode the contents of the file and inject it into a secret spec file and then use create/apply/delete as with other API objects but this just feels awkward because the create/apply/delete workflow is so clean.

For context this is a redis config file that contains a password and I'm mounting it with a volumeMount in a deployment:
```
- name: redis-fpc-conf-secret
  readOnly: true
  mountPath: /etc/redis
```

**Comment by eparis** [comment]
so you are hoping for `kubectl apply secret generic <name> --from-file=<key>=<file>` ?

**Comment by sijnc** [comment]
@eparis,

Something like this:
```
kubectl apply -f my-secret.yaml
```

my-secret.yaml
```
apiVersion: v1
	kind: Secret
	metadata:
	  name: my-secret
	dataFromFile:
	  key: redis.conf
```

When kubectl create or apply is run, kubectl would load and base64 encode the file on the client where kubectl is running before making its API calls.

**Comment by bgrant0607** [comment]
See also #19575, #22059

**Comment by bgrant0607** [comment]
@sijnc The API can't read your local file, so what you wrote won't work. 

Do you want the secret updated, or to create a new secret? Do you want all consumers of the secret to get the updated value simultaneously?

**Comment by bgrant0607** [comment]
cc @kubernetes/kubectl @pmorie

**Comment by bgrant0607** [comment]
You may be able to do:
```
kubectl create secret generic <name> --from-file=<key>=<file> --dry-run | kubectl apply -f -
```

**Comment by sijnc** [comment]
@bgrant0607 In my second comment I talk about kubectl loading the local file when running:

`kubectl create | apply -f my-secret.yaml`

When it encounters dataFromFile kubectl base64 encodes it before inserting it into the spec and submitting it to the API.

WTR Kubernetes updating the secret, it would probably be confusing if calling kubectl apply did not update the secret and it was consumed as a volume but how that happens, roll the pods or just update the file on the pod I haven't given much thought. I was more pointing out that it would be nice if the create/apply/delete workflow worked with secrets.

**Comment by pmorie** [comment]
@sijnc good news, @ingvagabund has made a PR today to make extend the secret volume and make it behave like the configmap volume: atomic updates when the secret updates, and control over filenames and subdirectories of keys in the volume

**Comment by pmorie** [comment]
@sijnc oops; https://github.com/kubernetes/kubernetes/pull/25285

**Comment by sijnc** [comment]
@pmorie #25285 ads KeyToPath and also atomic updates when the secret is updated. If you look at my first two comments they address using a secret with the create/apply/delete commands in kubectl.

I always found the base64 encoding of values in secret yaml files awkward and wondered why you can't just reference a path and kubectl could do the encoding before calling the kubernetes API.

It looks like the reasoning behind this is that the secret spec is an API object so adding dataFromFile to a spec that is ultimately consumed inside the cluster where the file is NOT present explains this and the decision to add the `--from-file=<file>` flag to kubectl was a better option.

Since this is a config file that is deployed as a secret I'll continue to keep the config file separate from the secret yaml file and just script the extra commands to make this work.

**Comment by pmorie** [comment]
@sijnc Yep, it was meant to be informational -- since the secret volume plugin doesn't update at all right now, it is a piece needed to give you what you want :)

I need to play around and see if I can make a recommendation to you on how to update the secret.  In the meantime, do check out the issue the @bgrant0607 linked (#22059) -- it would allow you to submit Secret descriptors with strings instead of base64 encoded byte arrays, which would effectively allow you to use `kubectl replace` (and perhaps `kubectl apply`, I will need to familiarize myself more with `apply`).

@liggitt any comment? ^

**Comment by sijnc** [comment]
@pmorie #22059 is a nice way of getting around the base64 legacy issue and allow raw strings, the way it's done could also be applied to a file by adding a second descriptor `file:path`.

Value is a string:
```
"username": "string:myuser"
```

Value is a file:
```
"config-file": "file:path/to/file.conf"
```

All good but you still have the issue where this needs to be handled on the client with the file and converted to string or base64 before submitting to the kubernetes API. The API vs kubectl issue wasn't fully apparent to me until @bgrant0607 pointed it out.

Adding special client side cases in the API spec are probably a bad idea but in the end I was just looking for a way to have a config file stored from a secret yaml that referenced the local config file but it's starting to look like the config file should be stored in a repo which can be pulled from the cluster, like the GitRepoVolumeSource with some type of authentication since this conf file contains a password.

Thanks for looking at this, let me know if you have any other thoughts.

**Comment by liggitt** [comment]
https://github.com/kubernetes/kubernetes/pull/22059 works server side as well. The API should not perform transformations/conversions that depend on state outside the API content (like an external file reference), in my opinion.

**Comment by sijnc** [comment]
@liggitt I agree and what I was suggesting happen client side would need to pollute the secret spec with an additional value that's not required in the API.

**Comment by bgrant0607** [comment]
Note that apply has the additional issue that secrets are captured by the config annotation: #23564.

**Comment by bgrant0607** [comment]
@sijnc It sounds like what you're looking for is client-side preprocessing. Some people in the community are doing that using various tools: jinja, jsonnet, etc. It would be easy to write a Makefile to translate from the raw secret data to a yaml file containing the K8s Secret, and then invoke `kubectl apply`.

**Comment by bgrant0607** [comment]
I expanded this to include configmaps, so that I could close #30337

**Comment by bgrant0607** [comment]
How I see this fitting into the big picture is covered by my uber doc: https://goo.gl/T66ZcD

**Comment by hofnarwillie** [comment]
This worked for me:

```
kubectl create secret generic <name> --from-file=<key>=<file> --dry-run -o json | kubectl apply -f -
```

**Comment by bgrant0607** [comment]
See also:
https://github.com/kubernetes-sigs/kustomize

**Comment by nyetwurk** [comment]
> This worked for me:
> 
> ```
> kubectl create secret generic <name> --from-file=<key>=<file> --dry-run -o json | kubectl apply -f -
> ```

I got this warning:
```
Warning: kubectl apply should be used on resource created by either kubectl create --save-config or kubectl apply
```

**Comment by fOO223Fr** [comment]
@nyetwurk you got this warning because when you first created the resource you didn't use --save-config flag [refer](https://kubernetes.io/docs/concepts/cluster-administration/manage-deployment/#in-place-updates-of-resources) here for detailed understanding of three way comparison which kubectl apply does

**Comment by nyetwurk** [comment]
@fOO223Fr Thanks!

**Comment by rdxmb** [comment]
> The drawback is this, If a change is made to the file you need to delete and then recreate the secret:
> 
> ```
> kubectl delete secret <name>
> kubectl create secret generic <name> --from-file=<key>=<file>
> ```

This seems also to be the case when not creating from file but creating from a CI-Pipeline like:

```
kubectl delete secret docker-registry
kubectl create secret docker-registry $DOCKER_REGISTRY --docker-server=$DOCKER_REGISTRY --docker-username=$DOCKER_REGISTRY_PULL_USER --docker-password=$DOCKER_REGISTRY_PULL_PASSWORD
```

**Comment by rikatz** [comment]
@sijnc is this issue still something? I guess during the discussion folks got a lot of different approaches :)

If not, I guess this can be closed right?
